libs/isograph-react/src/core/makeNetworkRequest.ts

```diff
@@ -44,6 +45,14 @@ export function maybeMakeNetworkRequest<
       return [wrapResolvedValue(undefined), () => {}];
     }
     case 'IfNecessary': {
+      if (
+        artifact.networkRequestInfo.normalizationAst.kind ===
+        'NormalizationAstLoader'
+      ) {
+        throw new Error(
```

Comment:
I love that if we have an entrypoint with an AST loader, we prevent `IfNecessary` at the type level for `useLazyReference`, can we do the same here?

That being said, (eventually?) you should be able to turn an `Entrypoint<..., NormalizationAstLoader>` into a `Promise<Entrypoint<..., NormalizationAst>>` and then load that second entrypoint with `IfNecessary`, if you choose to. So, you should be able to (with some extra work, and by being explicit about it) achieve what you're erroring out on

Reply:
I get pretty good (?) type errors with
```
export type FetchOptions<
  TReadOutData,
  NormalizationAstLoaded extends boolean,
> = {
  onComplete?: (data: TReadOutData) => void;
  onError?: () => void;
} & (NormalizationAstLoaded extends true
  ? {
      shouldFetch?: ShouldFetch;
    }
  : {
      shouldFetch: RequiredShouldFetch;
    });

export type IsographEntrypoint<
  TReadFromStore extends { parameters: object; data: object },
  TClientFieldValue,
  NormalizationAstLoaded extends boolean,
> = {
  readonly kind: 'Entrypoint';
  readonly networkRequestInfo: NetworkRequestInfo<
    NormalizationAstLoaded extends true
      ? NormalizationAst
      : NormalizationAstLoader
  >;
  readonly readerWithRefetchQueries: ReaderWithRefetchQueries<
    TReadFromStore,
    TClientFieldValue
  >;
  readonly concreteType: TypeName;
};
```

Reply:
Along with (with no overloads)
```
export function useLazyReference<
  TReadFromStore extends { parameters: object; data: object },
  TClientFieldValue,
  NormalizationAstLoaded extends boolean,
>(
  entrypoint: IsographEntrypoint<
    TReadFromStore,
    TClientFieldValue,
    NormalizationAstLoaded
  >,
  variables: ExtractParameters<TReadFromStore>,
  fetchOptions?: FetchOptions<TClientFieldValue, NormalizationAstLoaded>,
): {
  fragmentReference: FragmentReference<TReadFromStore, TClientFieldValue>;
} {
```

Reply:
Oh I guess we do need the overload then

Reply:
Okay, I can't obviously see a way to improve upon your approach. Maybe we might have cleaner type errors if we have two separate type names `IsographEntrypoint` and `IsographEntrypointWithNormalizationAstLoader`, but anyway, that is something we can play with after getting the functionality landed

Reply:
actually, I'm going back and forth. I do think it is nicer to have everything parametrized by a boolean (which could be "NormalizedAstPresent" | "AsyncNormalizationAst" or something for more clarity), even though we have to keep the overloads

I'm also not opposed to making the third arg not optional, as silly as it would look

Reply:
[check out this ts playground by lenz](https://www.typescriptlang.org/play/?ssl=166&ssc=1&pln=167&pc=1#code/JYWwDg9gTgLgBAbzgOQKYwO7QNYCVUCOArqgM4wCSAdgGYQA0cAogB4xQCGAxjAAoecQ6VFFKMAYpwDmQqjHw0RqKl1RwAvnBpQIIOAHIAAsFIQpnMAAsA9FFTcY+gLAAoVzACeYNcmggOADbAAF4cMMAQVACC5AAyEBwAJiJwALyIcAD6mYmRMNhUEBgwliae3gBccOwkGu5ePn6BIWER0eRpGdkQJSLlqFU1auquqCyQsNUNcADKlhBEAYni6FyWnfjEwHaJcwtLKzBrcAA+BhQ0aKqkpAIe+gDco+PQ8P1wm0TbqLvzi8urdbpfQATTI+lOBl8j1czwmb2mhzWAHkwOFIqQADwAFXwSWRRBgABEwhwAHydBCuOBwUh-A6AgD8VT2-yRlieLhpkQAwrowAFhMy4AAKRKkqq4+yJAnE0kASjSFIAbhBgIlOdyqEwoDooMKRYrUiq1RrXOpOXDXlNvB9CF8duzUeiqFipfjCSSYOTKdTafSAUdLFVPt9fvtA2tOZoAGRwJ1otpuvEyz2ksmWtwud5MORQLxquQARk6FFM5g4Vlz7ALwDkmL9SDAAg4QhgIlIVQgACMAFaoHgPODi71dvsD+Dqeh+gBEdAgM+nXJQTSCoRdMRgrjJrmSXACAjUXAx8GUNcgdZgRaq1fzF+LsOz01vtbkACZS+WLJYX-eYA3lybFs2w7Md+0HYcJTgHtwMnJcaTnCAF3glcoH8NdWkiTd4iSERt13AcDzsOBj1dU881fGA3xvCi-zfR9HzGeEbTUMszG-X9C3-P13WWHQQBmGBoDUMZ2yoRJSAyZtBGEUQwInIcRw4eSeA0FDsR5IIz3EYBUCWAA1QISHU3w0Oadc2k3bdfWXOwkkiAIPDgbA60SKp9E4y8YWXDhyElUz0JaDdyD9axrDgAA6KLzU5KllzCuAwnbcB4GvaDlREAIEgkv0aCIFQXTgIhSFQWIOGCDwFCUFRUAAmkaV4yRdEE4S4FE5QJKk4DZM7aDxwgpSVLgv0Gs03S5B0vTEkMgJjL9MkRRGtraK4qo2IrKsVsvOr6oalMmoEoS7BQ3aNK0ibdIMozUBO+qAvMzD2i3ZcaTJW7lQEYAOG7QVetYdgHH4GT21EHF9v4lq7DepbFCDZ0k2FBMXTdMbtMu6bruh5d5SqOLdu0DgZDPKq7BqqpJEJ2R5FQRRSdUMHpQOyGbrgM7xpgSartm1AMz9C1cvyng2iKkqyoqknlHppbGoho6RLYDrJKA4HQL62DFKgmCJzU6XUYuqaZrm5cFqWs871WuB1o4rb6yWvbGdl4TbtG86OfRw2brt1DAosrC4myvCXrgLHdo+qAvp+sgbzYTgeCB1seoZpImblkP6thlFEwxEN7TDJGkxxPW3YNzGUJxxAYekKmJbJ+Mq+JmnqvpmXmrlxg2bRkvud55d+eXPKCuF4rSvKyrG7p1BDQroO7BgIgoCoRBNF8xKqA8PmGOXYexbH2nJcns3KKLRgEHUeU-W30ea9UEVD7-Y+l5P-0I3ZKoZwuK4yFufMZw0c+sxpAlEoJg2q6mgLSIgXBsCkAvqLK+49963xtlRE+Z9QrhWAZJEQepaSoFQCASSEBsAcHXlvOB4sEE1SQeeLib5UFPzpC-QEb8P4Di-ncX+aCXAjCzHjOACUkr4LRHAaibUMr5mAVQKQcAQEcGqEQAUqABaD0iCLEeFC941R2vbZOjtiLtXEkrOA0kE4g16lrAamt+rDSDh3fWXMjavUWkHKKEUADad8uKMDDhHX6jAM6WHhhiAAulUNxptkFrS-JWH8yDtH1RbodJ2Xs7HFwcZ7IONJ7oYWCs9XaacaQ+O+r9aOAM47dTMUnPirdhIFK0ICIJrpEYNKzq6QurtOYY27rdYJkJwlB08ZeKJ7EYmeVtpk1m4ManHRSUXTpHtnbewerknCyQoBLTqUUyOf0Y6AwqR2KpKdam3QCY03qoZHQtORu09m8zS5LWCX6cufCaQEyJnIa+Aw66Uwbpo5uUyknHVZnM92mNNQaHBQPIWqjL4aKbpPRULy4Cz3novU+iVJIcDXhvLMsD1G73hdQ82l4H6n3-jSWFBKJ5EqPvQjIjC2TMLgO-S4bCbgcL-o+QBGDShYLAVACBUCYFkPxZ8mldFUHkv4TykB2DwElXwYQ4hpCKXkKpYgwZ746VIAZQyIMLDWXXG-h4Th-8eGuD4QImAyVhEAGZBiWDUBgeYgplo0MvDIySvQ4BSGUCIYAXBEpQCkEQKmyjoWL0pZ8+JYzTwK0MZbaJm13XjN2jSLFHhGD5QKEUKgWaqA5owFQDZ81nG7VcR45B3jPrFLIP4q5SZQnMGQW1eNnUrajLiV7OstNJkO2mRktNMjaApFSXc7miye0pCskHCkcBGSii9puVtYlOrZKCpZf2uEBULq9v0odNJNUwBoimmAizCk1u2aU2OfB9mg0SczOppzWnnNzpcuGL6bmd3SXUmkoS91e0PZE5tp7z1wC2SU5guzykq3vQCx9JyG0YmaR+65Y7QXdK9r0qVNIqhUFQOIp5uNK4-I+ZQ1Q5N65kb+bVB9bdgUdIwyQHuNI+6vMFoVKN5GEXT12iiheS8MWrxVXUXFIqd5iqPaSrhqrRXcfFVxaTDCAyv2Zawo1HKuHoOqLy0BODSCQOgXiiT8mj10KXlKoBum5UCoVQQ6CyrjPwJowpy85nT7KaYfqtThr2E-05dwoAA)

I haven't (yet) looked super closely into it, but perhaps we can improve on the error messages using some of the techniques there